retrieval/discovery/kubernetes.go

```diff
@@ -0,0 +1,582 @@
+	if err != nil {
+		// If we can't list nodes then we can't watch them. Assume this is a misconfiguration
+		// & panic rather than continue in a bad state
+		panic(fmt.Errorf("discovery.kubernetes.Sources: %s", err))
+	}
+	if res.StatusCode != http.StatusOK {
+		panic(fmt.Errorf("discovery.kubernetes.Sources: %s", res.Body))
+	}
+
+	var nodes kubernetes.NodeList
+	if err := json.NewDecoder(res.Body).Decode(&nodes); err != nil {
+		panic(fmt.Errorf("discovery.kubernetes.Sources: %s", err))
+	}
+
+	kd.nodesMu.Lock()
+	defer kd.nodesMu.Unlock()
+
+	sourceNames := make([]string, 0, len(nodes.Items))
+
+	kd.nodesResourceVersion = nodes.ResourceVersion
+	for idx, node := range nodes.Items {
+		sourceNames = append(sourceNames, kubernetesSourcePrefix+":"+node.ObjectMeta.Name)
+		kd.nodes[node.ObjectMeta.Name] = &nodes.Items[idx]
+	}
+
+	if kd.conf.DiscoverServices {
+		res, err := kd.client.Get(kd.conf.Server + kubernetesServicesUrl)
+		if err != nil {
+			// If we can't list services then we can't watch them. Assume this is a misconfiguration
+			// & panic rather than continue in a bad state
+			panic(fmt.Errorf("discovery.kubernetes.Sources: %s", err))
+		}
+		if res.StatusCode != http.StatusOK {
+			panic(fmt.Errorf("discovery.kubernetes.Sources: %s", res.Body))
+		}
+		var services kubernetes.ServiceList
+		if err := json.NewDecoder(res.Body).Decode(&services); err != nil {
+			panic(fmt.Errorf("discovery.kubernetes.Sources: %s", err))
+		}
+		if err != nil {
+			// If we can't list nodes then we can't watch them. Assume this is a misconfiguration
+			// & panic rather than continue in a bad state
+			panic(fmt.Errorf("discovery.kubernetes.Sources: %s", err))
+		}
+		kd.servicesMu.Lock()
+		defer kd.servicesMu.Unlock()
+
+		kd.servicesResourceVersion = services.ResourceVersion
+		for idx, service := range services.Items {
+			if kd.checkScrape(&service) {
+				sourceNames = append(sourceNames, kubernetesSourceServicePrefix+":"+service.ObjectMeta.Namespace+"/"+service.ObjectMeta.Name)
+				namespace, ok := kd.services[service.ObjectMeta.Namespace]
+				if !ok {
+					namespace = make(map[string]*kubernetes.Service)
+					kd.services[service.ObjectMeta.Namespace] = namespace
+				}
+				namespace[service.ObjectMeta.Name] = &services.Items[idx]
+			}
+		}
+	}
+
+	return sourceNames
+}
+
+func (kd *KubernetesDiscovery) checkScrape(s *kubernetes.Service) bool {
+	if scrape, ok := s.ObjectMeta.Annotations[kd.conf.ScrapeKey]; ok {
+		shouldScrape, err := strconv.ParseBool(scrape)
+		if err != nil {
+			log.Errorf("Invalid value (%s) for %s annotation on service %s/%s: %s", scrape, kd.conf.ScrapeKey, s.ObjectMeta.Namespace, s.ObjectMeta.Name, err)
+			return false
+		}
+		if shouldScrape {
+			return shouldScrape
+		}
+	}
+	return false
+}
+
+// Run implements the TargetProvider interface.
+func (kd *KubernetesDiscovery) Run(ch chan<- *config.TargetGroup) {
+	defer close(ch)
+
+	kd.updateNodesTargetGroup(ch)
+
+	for _, ns := range kd.services {
```

Comment:
So `Run()` is implemented based on the assumption that `Sources()` was called before?
The other implementations do not assume this, IIRC. But it is in fact how it works. I'm wondering whether it's worth adding this as an invariant to the interface.

Thoughts?